cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/BrokerSetAwareGoal.java

```diff
@@ -0,0 +1,320 @@
+                                                       .collect(Collectors.toSet());
+      if (maybeApplyBalancingAction(clusterModel, replica, eligibleBrokers, ActionType.INTER_BROKER_REPLICA_MOVEMENT, optimizedGoals,
+                                    optimizationOptions) == null) {
+        // If balancing action can not be applied, provide recommendation to add new Brokers.
+        ProvisionRecommendation recommendation =
+            new ProvisionRecommendation.Builder(ProvisionStatus.UNDER_PROVISIONED).numBrokers(clusterModel.maxReplicationFactor()).build();
+        throw new OptimizationFailureException(
+            String.format("[%s] Cannot move replica %s to %s on brokerSet %s", name(), replica, eligibleBrokers, originalBrokerBrokerSetId),
+            recommendation);
+      }
+    }
+  }
+
+  /**
+   * Check whether the given action is acceptable by this goal. The following actions are acceptable:
+   * <ul>
+   *   <li>All leadership moves</li>
+   *   <li>Replica moves that do not violate {@link #doesReplicaMoveViolateActionAcceptance(Replica, Broker)}</li>
+   *   <li>Swaps that do not violate {@link #doesReplicaMoveViolateActionAcceptance(Replica, Broker)}
+   *   in both direction</li>
+   * </ul>
+   *
+   * @param action       Action to be checked for acceptance.
+   * @param clusterModel The state of the cluster.
+   * @return {@link ActionAcceptance#ACCEPT} if the action is acceptable by this goal,
+   * {@link ActionAcceptance#BROKER_REJECT} if the action is rejected due to violating brokerSet awareness in the destination
+   * broker after moving source replica to destination broker, {@link ActionAcceptance#REPLICA_REJECT} otherwise.
+   */
+  @Override
+  public ActionAcceptance actionAcceptance(BalancingAction action, ClusterModel clusterModel) {
+    switch (action.balancingAction()) {
+      case LEADERSHIP_MOVEMENT:
+        return ACCEPT;
+      case INTER_BROKER_REPLICA_MOVEMENT:
+      case INTER_BROKER_REPLICA_SWAP:
+        if (doesReplicaMoveViolateActionAcceptance(clusterModel.broker(action.sourceBrokerId()).replica(action.topicPartition()),
+                                                   clusterModel.broker(action.destinationBrokerId()))) {
+          return BROKER_REJECT;
+        }
+        // SWAP Actions are performed by distribution goals to swap distinct partitions replicas to swap load in or out
+        // We would need to check both ways in swap cases that original brokerSet of both replicas matches the brokerSet of destination broker
+        if (action.balancingAction() == ActionType.INTER_BROKER_REPLICA_SWAP && doesReplicaMoveViolateActionAcceptance(
+            clusterModel.broker(action.destinationBrokerId()).replica(action.destinationTopicPartition()),
+            clusterModel.broker(action.sourceBrokerId()))) {
+          return REPLICA_REJECT;
+        }
+        return ACCEPT;
+      default:
```

Comment:
Shouldn't it also be ACCEPT if action is INTRA_BROKER_REPLICA_SWAP or INTRA_BROKER_REPLICA_MOVEMENT?

Reply:
I agree these should be accepted. Would you know why for RackAwareGoal these actions are not accepted either ?

[cruise-control/cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/AbstractRackAwareGoal.java](https://github.com/linkedin/cruise-control/blob/4e5927b48bf2581ab76acbbecbf42b355b871b65/cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/AbstractRackAwareGoal.java#L78)

Reply:
Hi @efeg , could you help us to understand why the switch doesn't contain all possible balance actions? E.g. INTRA_BROKER_REPLICA_SWAP and INTRA_BROKER_REPLICA_MOVEMENT are missing

Reply:
`INTRA_BROKER_REPLICA_SWAP` and `INTRA_BROKER_REPLICA_MOVEMENT` are relevant to goals configured with `intra.broker.goals`. This goal is an inter-broker goal, which is configured with the `goals` config. Inter and intra-broker goals do not work in a mixed mode -- i.e. their balancing actions are unsupported to work together.

Reply:
Does that mean CC will either run in intra-broker mode or inter-broker mode? Otherwise how does those different goals work together in the rebalance? (2-round goal optimization, 1 for inter-broker and 1 for intra-broker?)